### PR TITLE
Fix order of partial state tables when purging

### PR DESCRIPTION
**Pull request by DMRobertson**
See also https://github.com/matrix-org/complement/pull/608

`partial_state_rooms` has an FK on `events` pointing to the join event we get from `/send_join`, so we must delete from that table before deleting from `events`.

**NB:** It would be nice to cancel any resync processes for the room being purged. We do not do this at present. To do so reliably we'd need an internal HTTP "replication" endpoint, because the worker doing the resync process may be different to that handling the purge request.

The first time the resync process tries to write data after the deletion it will fail because we have deleted necessary data e.g. auth events. AFAICS it will not retry the resync, so the only downside to not cancelling the resync is a scary-looking traceback.

(This is presumably extremely race-sensitive.)

### CODE REVIEW

File: changelog.d/15068.bugfix [hunk 1]
```diff
@@ -0,0 +1 @@
+Fix a bug introduced in Synapse 1.76.0 where partially-joined rooms could not be deleted using the [purge room API](https://matrix-org.github.io/synapse/latest/admin_api/rooms.html#delete-room-api).
```
Comment (DMRobertson) [inline changelog.d/15068.bugfix:1]: This was (presumably) present in earlier Synapse versions too, if you had the experiemental faster joins feature turned on. But I didn't want to mention that here.

File: synapse/storage/databases/main/purge_events.py [hunk 2]
```diff
@@ -420,12 +420,14 @@ def _purge_room_txn(self, txn: LoggingTransaction, room_id: str) -> List[int]:
             "event_push_actions",
             "event_search",
             "event_failed_pull_attempts",
+            # Note: the partial state tables have foreign keys admist each other, and to
```
Comment (DMRobertson) [inline synapse/storage/databases/main/purge_events.py:423]:
```suggestion
            # Note: the partial state tables have foreign keys between each other, and to
```

**Comment by DMRobertson** [comment]
I don't know if we should consider this sufficient to close #14859 and #15067?

**Comment by DMRobertson** [comment]
Looks like I have complement tests to fixup.

**Comment by DMRobertson** [comment]
Arg. The complement test fails with workers.

File: synapse/storage/databases/main/purge_events.py [hunk 3]
```diff
@@ -420,12 +420,14 @@ def _purge_room_txn(self, txn: LoggingTransaction, room_id: str) -> List[int]:
             "event_push_actions",
             "event_search",
             "event_failed_pull_attempts",
+            # Note: the partial state tables have foreign keys between each other, and to
+            # `events` and `rooms`. We need to be delete from them in the right order.
```
Comment (squahtx) [inline synapse/storage/databases/main/purge_events.py:424]: small typo
```suggestion
            # `events` and `rooms`. We need to delete from them in the right order.
```

**Review by squahtx — APPROVED** [review]
lgtm!

**Pull request merged**